### DISCUSSION TITLE
Principled BSDF sampling

### DISCUSSION THREAD

**Discussion by PotatoPot6676** [mitsuba-renderer/mitsuba3#441] [Q&A] [answered]
Hi,

I am trying to understand how the principled BSDF sampling code works and I have some questions:

1. It appears that the sampling code takes into account, fresnel, when deciding to pick between reflection and transmission. I'm a bit confused as to how the probability of reflection (metallic and dielectric) in the following code `(1.0f - bsdf * (1.0f - F_spec_dielectric))` is calculated. Does it just mean Pr(reflection)= 1 - Pr(transmission)?

https://github.com/mitsuba-renderer/mitsuba3/blob/f3ac81bc5c6ce65d5843dde3a1d5f230353453e3/src/bsdfs/principled.cpp#L378-L387

2. When `spec_trans = 0` and `metallic = 0`, dielectric reflection and diffuse lobes are sampled equally (weights end up being 0.5 for both) and fresnel is not taken into account. Wouldn't sampling this way be inefficient to due the strong fresnel effect at grazing angles?

3. Would it make sense to take into account the luminance of the `base_color` while calculating the probability of sampling the diffuse lobe? Maybe it would provide slightly [better results](https://graphics.pixar.com/library/PxrMaterialsCourse2017/paper.pdf) when a dark `base_color` is used?

PS: My understanding maybe incorrect as I'm trying to figure things out :) Please feel free to correct me.

**Comment by njroussel** [comment]
Hi @PotatoPot6676 

For all of your questions, I would tend to agree. But I haven't looked deep enough into the implementation/literature to be fully confident in my answer.

@efylmzr off the top of your head, does this sound reasonable to you?
I can take a deeper look otherwise.

**Comment by efylmzr** [comment, marked as answer]
Hi @PotatoPot6676 
1- Yes, you are right, `Pr(reflection)= 1 - Pr(transmission)`. The sampling probabilities can not depend on Fresnel term in order to keep the normalization constant independent of Fresnel term.  Fresnel term is a value that we get after sampling the specular lobe. If you decide to sample the diffuse lobe, you do not have any Fresnel term to calculate the normalization constant. But if you both have a specular transmission and specular reflection lobe, you can decide to sample a halfway vector based on a probability which does not depend on the Fresnel coefficient. And then, you can utilize the Fresnel term to decide between transmission or reflection. So sum of selection probabilities of specular reflection and transmission must not depend on Fresnel term. This is why I did not utilize the Fresnel coefficient of clearcoat lobe while sampling, because there is no transmission lobe.  
2- Yes you are right, the problem here is the same as the first answer. We choose a lobe based on the parameters of the model and sample it afterwards. Fresnel term is a value that we get after sampling a halfway vector. To understand why it is a problem, think about the pdf calculation. While we are calculating the pdf, we just have `wi` and `wo` as input. Then we calculate the probability of each lobe based on these 2 directions and add them up with sampling weights. If you want to calculate the pdf of diffuse lobe, you can not know the halfway vector because you are assuming that `wo` is sampled with cosine hemisphere. The only way to utilize the Fresnel term in sampling is to calculate it based on the incoming direction and the surface normal instead of the microfacet normal. But then, you are not sampling proportional to the bsdf function.
3- Yeah it may be. I thought about this before. But then I did not include it, I do not remember the exact reason but the difference is not that much if you do not use a complete black color. Also, the Fresnel term is affecting the result of the BSDF which we do not take into account while sampling. There is a sheen lobe which does not depend on the color and specular reflection is not suitable for sampling it so you can not turn of the diffuse sampling even though the color is black.  It is a bit hard to find a heuristic to make the sampling perfect.  
Please feel free to ask if it is not clear.

**Comment by PotatoPot6676** [comment]
Thank you very much @njroussel and @efylmzr for the very detailed answer. Your explanation about the fact that the normalization constant should be independent of the Fresnel term helped me fix my lobe sampling probabilities :) I have some follow up questions though:

How did you validate that the sampling probabilities that you used get the correct results? I went back and validated my implementation by setting all lobes to the same probability and manually comparing results against Fresnel weighted probabilities. Is there a better way to test this?

Also, a question not related to sampling but the clearcoat evaluation:

I'm unable to figure out where `dr::abs(cos_theta_o)` in the following evaluation code for clearcoat comes from. Clearcoat uses the same Smith G derivation (with fixed roughness) as the regular reflection lobe so should it not be `1.0 / (4.0f * dr::abs(cos_theta_i))` similar to the regular reflection evaluation? The $\Large \frac{1}{4cos\theta_{i}}$ factor from the equation $\Large \frac{D G F}{4cos\theta_{i}cos\theta_{o}}$ does not seem to get cancelled in F, D or G but the cosine foreshortening factor $\{cos\theta_{o}}$ should get cancelled out. Perhaps I have missed something but I get somewhat darker results at glancing angles with the current Mitsuba code which made me wonder.

Clearcoat:
https://github.com/mitsuba-renderer/mitsuba3/blob/8ebf3af3021ee53087b15af51361a3339655c3ab/src/bsdfs/principled.cpp#L640

Specular Reflection:
https://github.com/mitsuba-renderer/mitsuba3/blob/8ebf3af3021ee53087b15af51361a3339655c3ab/src/bsdfs/principled.cpp#L601

**Comment by efylmzr** [comment]
Hi, @PotatoPot6676 ,
I used Chi2 test to validate the sampling and pdf functions.
Cosine terms are coming from the definition of the rendering equation. Theta_o is the direction that goes to light, not the camera.  It corrects the radiance value coming from the outgoing radiance. For clearcoat lobe, Disney course notes did not define (1/4 cos(i) cos(o)) term. So the bsdf function of clearcoat is only (F * D * G).

**Comment by PotatoPot6676** [comment]
@efylmzr: Ah I see. I had assumed clearcoat to include the $\Large \frac{1}{4cos\theta_{i}cos\theta_{o}}$ factor. The code makes sense now. 

Thanks a lot for taking time out to respond and for all your help :)
